File: app/src/main/java/com/esorokin/lantector/presentation/presenter/BasePresenter.java
package com.esorokin.lantector.presentation.presenter;

import com.arellomobile.mvp.MvpPresenter;
import com.arellomobile.mvp.MvpView;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public abstract class BasePresenter<V extends MvpView> extends MvpPresenter<V> {
	private final CompositeDisposable compositeDisposable;

	public BasePresenter() {
		super();
		this.compositeDisposable = new CompositeDisposable();
	}

	protected void autoDispose(Disposable disposable) {
		compositeDisposable.add(disposable);
	}

	@Override
	public void onDestroy() {
		super.onDestroy();
		compositeDisposable.dispose();
	}
}

File: app/src/main/java/com/esorokin/lantector/presentation/presenter/BasePresenter.kt
package com.esorokin.lantector.presentation.presenter

import com.arellomobile.mvp.MvpPresenter
import com.arellomobile.mvp.MvpView
import com.art.alligator.Navigator
import com.esorokin.lantector.utils.ext.AutoDisposableContainer
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import javax.inject.Inject

abstract class BasePresenter<V : MvpView> : MvpPresenter<V>(), AutoDisposableContainer {
    @Inject
    internal lateinit var navigator: Navigator

    private val compositeDisposable: CompositeDisposable = CompositeDisposable()

    override fun autoDispose(disposable: Disposable) {
        compositeDisposable.add(disposable)
    }

    override fun onDestroy() {
        super.onDestroy()
        compositeDisposable.dispose()
    }
}